### ISSUE TITLE
Thread safety

### ISSUE THREAD

**Issue by chansen** [p5pclub/http-xsheaders#1]
It's probably a good idea to add a notation in documentation about thread safety, current implementation of HTTP::XSHeaders is not thread-safe.

--
chansen

**Comment by chansen** [comment]
Either document the limitation or use MAGIC and svt_dup hook to make HTTP::XSHeaders thread safe.

**Comment by xsawyerx** [comment]
@chansen I know "patches welcome" on this is pretty broad, but both @gonzus and I have less familiarity with making XS thread-safe and would appreciate help, if you have any tuits available.

**Comment by chansen** [comment]
@xsawyerx, we work together to make CPAN a better place =)! I'll take a deeper look at it later this week and provide a pull request if time permits, otherwise it will happen next week.

**Comment by xsawyerx** [comment]
Tomorrow we're sitting to code again, so we'll try and handle it. Hopefully we succeed. Will update!

**Comment by chansen** [comment]
My fork of [HTTP::Headers](https://github.com/chansen/http-xsheaders/tree/typemap) adds thread safety to the current implementation!  We should discuss performance issues in a different issue/thread. 

--
chansen

**Comment by gonzus** [comment]
Hi chansen,

This is awesome. I would love to include your work in the original version; can we do that? Pull request?

Please open a new issue for the performance issues you may have observed, so that we can discuss them. Keep in mind that the first implementation of the C data structures is very simple minded, and works well for small numbers of headers (and values in them), which we theorised would be the typical case.

Thanks!

**Comment by chansen** [comment]
Hi @gonzus,

I'm sorry for the late response, I have had a flu. It was my intention that my fork should be integrated in the original version. I saw on CPAN that @xsawyerx has released v0.400000, with my changes incorporated, so it's probably a bit late for me to submit a pull request.

I'll open new tickets for suggested performance improvements.

**Comment by xsawyerx** [comment]
@chansen I'm sorry, we can be terrible communicators at times. @gonzus and I were sitting down during an AmsterdamX Perl Mongers meeting and Gonzalo noted we have to merge your stuff and asked if we can do it without you having to open a pull request for it. So we just merged it and released it on the spot.

We appreciate your help and we'd like to apologize for not communicating that we've merged and released those changes.

Have a speedy flu recovery! :)
